### PR TITLE
New algorithm to group point data to X resolution

### LINKED ISSUE #23908: Extract grouping from ReflectometryILLConvertToQ to a separate algorithm
The grouping process in `ReflectometryILLConvertToQ._groupPoints()` should be moved to a real algorithm to facilitate testing, improve performance and generally move `ReflectometryILLConvertToQ` towards being a real WorkflowAlgorithm without it's own processing methods.

### PR DESCRIPTION
**Pull request by soininen**
This PR introduces a new algorithm `GroupToXResolution` which groups points within a given fraction of Dx from each other into single points. The algorithm is meant to regroup reflectivity data where the Dx values correspond to the Qz resolution and it is useless to have a large number of points within the resolution distance from each other. The implementation originates from the legacy LAMP/COSMOS program.

As a small bonus this PR adds a new `WorkspaceHasDxValidator`.

**Report to:** [nobody]

**To test:**

The algorithm can be tried using the following script:
```python
import numpy as np
# Create a workspace with exponential decay.
Xs = np.arange(0.01, 5., 0.01)
Ys = np.exp(-Xs)
DXs = np.empty_like(Ys)
DXs.fill(0.1)
original = CreateWorkspace(Xs, Ys, Dx=DXs, NSpec=1)
grouped = GroupToXResolution(original, FractionOfDx=1.)
```
Check that with the Dx set to 0.1 and `FractionOfDx` to 1.0, the X separation of points in the `grouped` workspace should be around 0.1. Plot the two workspaces to see that the data matches. Check that the errors (both E and Dx) are sound. Try changing the `FractionOfDx` or Dx and check that the results change accordingly.

Fixes #23908.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/test/GroupToXResolutionTest.h [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    TS_ASSERT_EQUALS(outputWS->blocksize(), 1)
+    TS_ASSERT_EQUALS(outputWS->x(0).front(), 0.23)
+    TS_ASSERT_EQUALS(outputWS->y(0).front(), 1.42)
+    TS_ASSERT_EQUALS(outputWS->e(0).front(), std::sqrt(1.42))
+    TS_ASSERT(outputWS->hasDx(0))
+    TS_ASSERT_EQUALS(outputWS->dx(0).front(), 1.)
+  }
+
+  void test_two_points_get_averaged() {
+    using boost::math::pow;
```
Comment (Pasarus) [inline Framework/Algorithms/test/GroupToXResolutionTest.h]: Can this just be moved out to the top of the file?

**Review by Pasarus — CHANGES_REQUESTED** [review]
The XError data on the new workspaces have different values depending on what column, unlike the original workspace, is this correct?:
image

Other than the noted things, it looks good :+1:

**Comment by soininen** [comment]
@Pasarus The larger XError is expected as the Dx for the grouped point is calculated as

```
Dx_new = sqrt(Dx_first^2 + (0.68 * (X_last - X_first))^2),
```

where `Dx_first` is the Dx of the first point to group and `X_last - X_first` is the X range (span) of all points in the group.

**Comment by Pasarus** [comment]
Yep that makes sense thanks for clearing that up :+1:

File: Framework/Algorithms/src/GroupToXResolution.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  outXs.reserve(inXs.size());
+  std::vector<double> outYs;
+  outYs.reserve(inXs.size());
+  std::vector<double> outEs;
+  outEs.reserve(inXs.size());
+  std::vector<double> outDxs;
+  outDxs.reserve(inXs.size());
+  size_t pointIndex{0};
+  double begin = inXs.front();
+  while (true) {
```
[the remark this answers is not in the document]
Comment (soininen) [reply Framework/Algorithms/src/GroupToXResolution.cpp]: Makes sense as that would be the maximum possible size of the grouped histograms as well.

[the remark this answers is not in the document]
Comment (soininen) [reply Framework/Algorithms/src/GroupToXResolution.cpp]: I ended up keeping the `while (true)` but throwing if the next step would be backwards or nowhere at all.

File: Framework/Algorithms/test/GroupToXResolutionTest.h (hunk 1 above)
Comment (soininen) [reply Framework/Algorithms/test/GroupToXResolutionTest.h]: Personally, I like it a bit more here, but I'll move it anyway.

**Review by Pasarus — DISMISSED** [review]
Looks good to me :+1:

**Comment by soininen** [comment]
@Pasarus Resolving the merge conflicts with master got your review dismissed. Could you re-approve? Only the `reflectometry.rst` release notes were modified to resolve the conflicts.

**Review by Pasarus — DISMISSED** [review]

File: Framework/Algorithms/src/GroupToXResolution.cpp [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+#include <boost/math/special_functions/pow.hpp>
+
+namespace {
+namespace Prop {
+std::string const FRACTION{"FractionOfDx"};
+std::string const INPUT_WS{"InputWorkspace"};
+std::string const OUTPUT_WS{"OutputWorkspace"};
+} // namespace Prop
+constexpr double FWHM_GAUSSIAN_EQUIVALENT{0.68};
```
Comment (gvardany) [inline Framework/Algorithms/src/GroupToXResolution.cpp]: Would it make sense to have this as another input in multiples of sigma?

File: Framework/Algorithms/src/GroupToXResolution.cpp [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  declareProperty(Prop::FRACTION, 0.2, betweenZeroAndOne,
+                  "A fraction of Dx to group the points to.");
+}
+
+std::map<std::string, std::string> GroupToXResolution::validateInputs() {
+  std::map<std::string, std::string> issues;
+  API::MatrixWorkspace_const_sptr inWS = getProperty(Prop::INPUT_WS);
+  if (inWS->getNumberHistograms() != 1) {
+    issues[Prop::INPUT_WS] =
+        "The workspace should contain only a single histogram.";
```
Comment (gvardany) [inline Framework/Algorithms/src/GroupToXResolution.cpp]: Is this not too strict? One can in principle have ragged output.

File: Framework/Algorithms/src/GroupToXResolution.cpp [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      boost::make_shared<API::HistogramValidator>(acceptHistograms));
+  declareProperty(
+      Kernel::make_unique<API::WorkspaceProperty<>>(
+          Prop::INPUT_WS, "", Kernel::Direction::Input, inputValidator),
+      "An input workspace with Dx values.");
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::OUTPUT_WS, "", Kernel::Direction::Output),
+                  "The grouped workspace.");
+  auto betweenZeroAndOne =
+      boost::make_shared<Kernel::BoundedValidator<double>>(0., 1.);
```
Comment (gvardany) [inline Framework/Algorithms/src/GroupToXResolution.cpp]: I don't see why this has to be capped by 1. One might want to group based on multiples of the resolution.

File: Framework/Algorithms/src/GroupToXResolution.cpp (hunk 3 above)
Comment (soininen) [reply Framework/Algorithms/src/GroupToXResolution.cpp]: It is from COSMOS. I think it has some statistical backing and should not be modifiable by the user.

File: Framework/Algorithms/src/GroupToXResolution.cpp (hunk 4 above)
Comment (soininen) [reply Framework/Algorithms/src/GroupToXResolution.cpp]: We can expand the algorithm later if the need arises. At this moment, a single histogram treatment suffices for the foreseen use cases.

**Comment by gvardany** [comment]
Also do you think a performance test would be useful?

**Comment by gvardany** [comment]
@Pasarus Do you mind checking the changes and approving again?

**Comment by Pasarus** [comment]
@gvardany Can do :+1:

**Review by Pasarus — CHANGES_REQUESTED** [review]
I wonder whether it's needed to do 5000 iterations for the performance test, because it takes 37 seconds for the test to run by itself, perhaps 500 iterations would be sufficient?

**Comment by soininen** [comment]
@Pasarus Did you use an optimised release build? On my machine the test runs in ~5 seconds.

**Comment by Pasarus** [comment]
I did not, I ran it on a developer debug build, are the build servers/developers running the tests going to be running on an optimized release build?

**Comment by soininen** [comment]
Yes, it does not make sense to test performance on unoptimized builds.

**Review by Pasarus — APPROVED** [review]
Had to make a release optimized build, changes are good :+1:

**Pull request merged**